src/shims/unix/wasi/foreign_items.rs

```diff
@@ -0,0 +1,187 @@
+use std::ffi::OsString;
+
+use rustc_span::Symbol;
+use rustc_target::abi::Size;
+use rustc_target::spec::abi::Abi;
+
+use crate::*;
+use shims::foreign_items::EmulateByNameResult;
+
+use rustc_middle::ty::layout::LayoutOf as _;
+use rustc_middle::ty::Ty;
+
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriInterpCx<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
+    fn emulate_foreign_item_by_name(
+        &mut self,
+        link_name: Symbol,
+        abi: Abi,
+        args: &[OpTy<'tcx, Provenance>],
+        dest: &PlaceTy<'tcx, Provenance>,
+    ) -> InterpResult<'tcx, EmulateByNameResult<'mir, 'tcx>> {
+        let this = self.eval_context_mut();
+
+        // See `fn emulate_foreign_item_by_name` in `shims/foreign_items.rs` for the general pattern.
+        match link_name.as_str() {
+            "args_sizes_get" => {
+                let [num_args_ptr, cumulative_size_ptr] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let ptr_usize = this.tcx.mk_mut_ptr(this.tcx.types.usize);
+
+                let num_args_ptr = this.addr_to_mplace_ty(num_args_ptr, ptr_usize)?;
+                let cumulative_size_ptr = this.addr_to_mplace_ty(cumulative_size_ptr, ptr_usize)?;
+
+                let argc: u32 = this.machine.args.len().try_into().unwrap();
+                let cumulative_size: u32 = this
+                    .machine
+                    .args
+                    .iter()
+                    .map(|e| {
+                        // Include null terminator
+                        e.as_bytes().len().checked_add(1).unwrap()
+                    })
+                    .sum::<usize>()
+                    .try_into()
+                    .unwrap();
+
+                // Store argc and cumulative size into the pointers passed in as arguments.
+                this.write_scalar(Scalar::from_u32(argc), &num_args_ptr.into())?;
+                this.write_scalar(Scalar::from_u32(cumulative_size), &cumulative_size_ptr.into())?;
+
+                this.write_scalar(Scalar::from_u32(0), dest)?;
+            }
+
+            "args_get" => {
+                let [argv, argv_buf] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                this.args_get(argv, argv_buf)?;
+
+                this.write_scalar(Scalar::from_u32(0), dest)?;
+            }
+
+            "fd_write" => {
```

Comment:
this seems complex enough to deserve a method. Also leave a comment why this is needed and can't go through the regular file operations.